### ISSUE TITLE
Support Fastify

### ISSUE THREAD

**Issue by llong2195** [nestjs-shopify/nestjs-shopify#202]
Nestjs supports both Express and Fastify. Currently, it only supports Express. Do you have plans to support Fastify in the future?

**Comment by tolgap** [comment]
I think we should be able to figure out `fastify` usage based on the `HttpAdapter.getType()` call and check if it equals `'fastify'`: https://github.com/nestjs/nest/blob/master/packages/platform-fastify/adapters/fastify-adapter.ts#L559-L561

If this is the case, we should use `req.rawRequest`.

**Comment by tolgap** [comment]
@llong2195 sure! It's open source on Github, just so everyone can contribute!

Places where the request and response are directly used:

* https://github.com/nestjs-shopify/nestjs-shopify/blob/main/packages/auth/src/auth.filter.ts#L26..L27
* https://github.com/nestjs-shopify/nestjs-shopify/blob/main/packages/auth/src/auth.guard.ts#L74..L75
* https://github.com/nestjs-shopify/nestjs-shopify/blob/main/packages/auth/src/auth-base.controller.ts#L22..L23
* https://github.com/nestjs-shopify/nestjs-shopify/blob/main/packages/auth/src/auth-base.controller.ts#L47..L48
* https://github.com/nestjs-shopify/nestjs-shopify/blob/main/packages/graphql/src/graphql-proxy/graphql-proxy.controller.ts#L18..L19

Basically, it would be nice to have a utility function where you can pass a request of type `any`, and return the correct `IncomingMessage` from that request.

```ts
function getRawRequest(req: any, httpAdapter: HttpAdapter) {
  if (httpAdapter.getType() === 'fastify) {
    return req.raw
  }
  
  return req
}

**Comment by bertiepinnock** [comment]
Hi @tolgap - I was wondering if you had any idea when #309 might be merged/released? We would love to be able to use this set of packages with our NestJS services, but unfortunately my organisation relies quite heavily on Fastify so currently it's incompatible.

I'd be happy to contribute/help out if needed, but it looks like #309 is already a complete implementation!

**Comment by tolgap** [comment]
@bertiepinnock @llong2195 @schiller-manuel I am tagging all you along for thoughts 🙂 .

I have been apprehensive of merging any of my own `fastify` PRs because I think it goes against the "NestJS way™" of doing things. Usually what NestJS promotes is splitting up `express` stuff into their own package, and `fastify` stuff into their own package.

But that would significantly change the API of the 4 existing packages 😞. Basically, all of `@nestjs-shopify/auth` would completely depend on either `@nestjs-shopify/express` or `@nestjs-shopify/fastify`.

My question is: do you think its splitting this repository further up to support an `express` and `fastify` specific package? It would definitely be a hassle to upgrade for any users of the current `@nestjs-shopify/auth` package.

PS: I'm pinning this issue for discussion!

**Comment by tolgap** [comment]
@schiller-manuel yes using an adapter pattern we should be able to reduce the amount of API changes necessary for the upgrade path to `express` and `fastify` specific packages 👍 .

**Comment by tolgap** [comment]
@schiller-manuel @bertiepinnock @llong2195 we now have native `fastify` support! Please see the upgrade doc on how to migrate to it! https://github.com/nestjs-shopify/nestjs-shopify/blob/main/docs/migrate-to-express-or-fastify-package.md

**Comment by bertiepinnock** [comment]
That's great, thank you @tolgap 🙌
